### ISSUE TITLE
font test string: leverage kerning

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP#216]
something that has bugged me for a while

I have improved my string ~~to detect more fonts that the current (by 1 so far)~~ to increased the number of size buckets. Currently this is just using plain latin chars (no accents etc). We don't really want to get into other chars AFAICT, but I have a 4000 char pairing list that I will check out.

Will be dumping notes here as I go

**Comment by Thorin-Oakenpants** [comment]
https://en.wikipedia.org/wiki/Kerning#Kerning_values

ignoring the values listed in the link, as fonts will do whatever they want - this is just to get common pairings

original list to start concatening some examples to reduce chars used - e.g CO + OC = COC
```
VA
YA
Te
AV
Vr
PA
FA
UA
Yt
LT
Xv
Ku
OA
Hv
T:
DY
my
Ru
aj
bv
Sp
ro
SR
lp
ot
tt
am
fe
vo
xc
yo
Ix
st
he
Fw
us
Ak
la
Oj
il
CO
bc
Xf
fr
wb
YW
So
Co
VT
cv
Dv
OC
Bc
RX
gy
XA
ry

w.
r,
e,
f?
r:
w;
D,
W.
P,
F.

A\"
D\"
F\"
L\"
T\”
a\"
c\"
f\”
m\"

```

**Comment by Thorin-Oakenpants** [comment]
Actually the longer the string...

you can end up reducing size buckets by adding more chars - so technically you want to test all pairs with a space between them, so it only adds differences

**Comment by Thorin-Oakenpants** [comment]
First things first: I can't improve font detection: basically all you need is a simple string and a large enough size to differentiate from the base font. So all I can do is increase size buckets

I worked my way thru 4000+ char pairs from a font kerning list, which I joined up into long strings like this

```
nto)rzwtyxyšçeýsýšňtŠuΕ,κινωВ,В.ато)оух,х.цещею)(YC,DULMLlX.ZTbæ
fufúpæs.vivpvívîwiwpš.Φοη,ωπгуддзъсяцо(bBZMiMít»uyvuvüwuwüyiüyοπ
ωνАсИнШпЯкоеооософслфухлэсэфюс(б(з(э-wEMPDPEPFPHPIPKPLPPPRYVYWat
gmgngrkbq.rurúrûrüátâtätåtΕκορυπБлЖиКиСхШвЯвозрууеюбюз(v-zD’fafg
fofáfðfóføkikíyfÀsÄsÅsνητηсаCYRMj,vfëzězα-βρАлзусстхукунупуюх:х;
AMLjvtw-ΑιναρνρπυνχεЛасжт)ф,’S.tLJScSeSoSóSöSøezgigìgíx,ypýpŠeСе
```

And then I added each line, one by one, to the fontStr (the long one from above) to see if it increased buckets. Sometimes the number of size buckets would decrease - and I worked out that some chars cause this loss. So I would bisect to identify it, then remove it from the lines - this left some single chars without a pair in the lines, big deal - all good pairs were still in there.

And when size buckets increased, I would bisect to find what added them

I also bumped the size from 256px to 512 (it added one more size bucket)

Anyway, from 240 windows fonts (of which my system has 143)
```js
original = "mmmLLLmmmWWWwwwmmmllliii", // 116 size buckets
foryourinfo = "mLWwli" // 117

fntStrA = "OCOAVTe,SRXfrottRustamyoilajBcvoIxc\"wbcXAkLT\"DYtYW.Vr,Fw;PA\"Co"
	+ "FAKuDvD,D\"F.F\"HvL\"OjP,SoSpT:UAVAXvYAa\"bvfef\”f?gyhelpm\"ryr:w." // 123
fntStrA += "à" // 125
fntStrA += "к" // 127
fntStrA += "-î-" // 128

```

Notice the last three additions contain cryllic or accents

**Comment by Thorin-Oakenpants** [comment]
Now one thing I have noticed is that the font test has increased from approx 65 to 90ms. I'm not sure why this is _exactly_ - number of chars, chars not supported in that font (e.g. those last three), or the variety of chars

One part is that I have increased the fonts tested from 218 to 240 - so that accounts for 6ms or so. But the other 20ms needs investigation.

Is it really that important to get unique sizes per font vs perf? Probably not. I have been using common sizes to remove some fonts from the font list, but I think we should keep them all instead (users can add/remove fonts) - except for a very few (such as courier = courier new, time = times new roman, etc.

I don't think the absolute max precision and as many sizes buckets as possible is that important

Also, this is an imperfect test. It's only on one machine (at DPI/devicePixelRatio 1 at normal scaling etc), and doesn't cover all the fonts we want to test,

To be continued

**Comment by Thorin-Oakenpants** [comment]
these are practically identical in terms of perf
```
fntStrA = "mmmLLLmmmWWWwwwmmmllliii", // 116
fntStrA = "mLWwli" // 117
```

at 123 buckets vs 128 buckets (the cyrillic and accented chars), the diff is about 3 to 5 ms
```
fntStrA = "OCOAVTe,SRXfrottRustamyoilajBcvoIxc\"wbcXAkLT\"DYtYW.Vr,Fw;PA\"Co"
	+ "FAKuDvD,D\"F.F\"HvL\"OjP,SoSpT:UAVAXvYAa\"bvfef\”f?gyhelpm\"ryr:w."
 // 123 buckets | perf: 95, 91, 91, 91, 91, 92 (section reruns) - avg 91.8


fntStrA += "à" // 125
fntStrA += "к" // 127
fntStrA += "-î-" // 128 buckets | perf: 95, 93, 94, 94, 92, 95 (section reruns) avg 93.8
```

the super tiny string with accents is 125 which looks promising (but IDK about other OSes)
```
fntStrA = "mLWwli" // 117 buckets | perf: 74, 74, 73, 72, 73, 70 (section reruns) avg 72.7
fntStrA += "à" // 120
fntStrA += "к" // 125 buckets | perf: 72, 73, 71, 72, 72, 70 (section reruns) avg 71.7
```

I think we can say it is simply the variety of chars being looked up that causes this. So a happy fast string that gets some decent size buckets I guess (and to be tested on other platforms and against more fonts)

"mLWwli" was made up by me as fonts that are wide and thin and would show up diffs easily. It's going to take some experimentation to find the perfect string

**Comment by Thorin-Oakenpants** [comment]
```
let fntStrA = "CWàкiW\”"
/* "CW" 117 + "à" 121 + "к" 125 + "i" 126 + "W\”" 127 (win 7)
```

I'll go with this which is 6 unique chars used compared to the original seven. I can always tweak it per OS but I would think it should work well everywhere

**Comment by Thorin-Oakenpants** [comment]
> "mLWwli" was made up by me as fonts that are wide and thin

wide and _narrow_ rather

which is what bugged me - the fatness and narrowness are contradictory

anyway, closing - I can test the string on macOS and linux when I get around to them. You can just do it from the console and click rerun

**Comment by Thorin-Oakenpants** [comment]
definitely going to use custom strings per OS and TB vs FF

Windows TB gets the best results with `àк”,?-` - the notos love the last three chars .. so my default is going to be `CWàкiW\”,?-` and I'll tweak it as I get to each OS

**Comment by Thorin-Oakenpants** [comment]
this is local code only, but I may make a poc for it - at the moment it is part of TZP and simply calls the get_fonts() function and records the number of size buckets

note with steps 1 the function is using widely super supported latin, cyrillic, greek chars - i.e not going to be tofu. in step four we are using zero-width non-joiners and covering a large range of code points (from my font script PoC) the default array is 6k+ chars but excludes LTR languages. The remainder is another 6.9k chars. The parameters are (baseString, useCodePoints, source) - useCodePoints is there because you can also use arrays of chars in this function. Anyway, so far step four is not producing anything not supported (except in TB where tofu causes almost every font to be detected) - the key is you can look at the results and eliminate tofu and make a decision on a char's support

here's windows 7: the resultant 4 letter string cannot be improved upon and makes for a good start to see if we can improve in win10. I will post windows TB results soon - windows TB is awesome because the fonts are 99% bundled, but ... but a bunch of notos have no size entropy - still, so many scripts covered that it made it easy

----

step 1, create pairs
step 2, test
step 3, get the top pairs

![step1-3](https://user-images.githubusercontent.com/16656956/208715828-87c92388-fd00-40bf-a6cd-02abdca0bbd4.png)

---

repeat steps 1 to 3 but pass the top pairs as an array

![step1-3-repeat](https://user-images.githubusercontent.com/16656956/208715945-34def1d5-2cfa-46b7-ab1a-e570a1b64fbe.png)

---

step 4, test the trips against a bunch of unicode code points

I went with `Mōá` for step 4

Here's step 4 in action
![step4](https://user-images.githubusercontent.com/16656956/208716088-819f7f1e-6e3d-4db3-8acb-06ec231795ed.gif)

step 4 finished
![step4](https://user-images.githubusercontent.com/16656956/208716154-31dda3fc-79fc-4b1f-88a4-5b5fef3c661b.png)

That's it .. `MōáΩ` - 4 chars, 127 size buckets - it cannot be improved on at least on windows 7: what size collisions there are in detected fonts are obvious such as aliases or they were designed that way
- "MS Shell Dlg", "Microsoft Sans Serif"
- "MS Gothic", "MS Mincho", "ＭＳ ゴシック", "ＭＳ 明朝"
- "NSimSun", "宋体"
- "Microsoft YaHei", "微软雅黑"
- "MingLiU", "MingLiU_HKSCS"
- "Cordia New", "CordiaUPC"
- "Browallia New", "BrowalliaUPC"
- "Angsana New", "AngsanaUPC"

**Comment by Thorin-Oakenpants** [comment]
> Anyway, so far step four is not producing anything not supported (except in TB where tofu causes almost every font to be detected)

OK, so the tofu is fine (just use code point `0xFFFF` to test). Detected fonts doesn't change, it's the size buckets that do - dramatically which is why I thought initially they were an issue. 

Adding a tofu for TB on windows jumps the size buckets from 89 to 132. I can eke that to 133 with another char but it's not worth it This should also work on mac + linux because all three platforms have 118 noto fonts in common - but will test

So the best pairs/trips are combos of `,`, `-`, and `?`. Add a tofu and we can drop `,`. So the final windows TB fntString is `?-` + a tofu (0xFFFF). Note I tested a tofu on FF windows and no effect, so it's clearly a noto thing. I also tested TB windows with a full fntList of the allowlist (TB's whitelist) and all 240 windows system fonts - same result, so no false positives.

Wow: windows(7)
TB: 3 chars, 132 buckets
FF: 4 chars, 127 buckets

Seems like windows likes a little mix of some AZ Latin capitals, a little small cyrillic and a touch of greek. And for TB bundled notos replace so many other fonts that some punctuation chars and a tofu kicks ass
